tools/resourcedocsgen/cmd/docs/registry.go

```diff
@@ -126,31 +129,71 @@ func genResourceDocsForAllRegistryPackages(registryRepoPath, baseDocsOutDir, bas
 		return errors.Wrap(err, "reading the registry packages dir")
 	}
 
-	for _, f := range metadataFiles {
-		glog.Infof("=== starting %s ===\n", f.Name())
-		glog.Infoln("Processing metadata file")
-		metadataFilePath := filepath.Join(registryPackagesPath, f.Name())
+	var m sync.Mutex
+	var errs []error
+	addError := func(err error) {
+		m.Lock()
+		defer m.Unlock()
+		errs = append(errs, err)
+	}
 
-		b, err := os.ReadFile(metadataFilePath)
-		if err != nil {
-			return errors.Wrapf(err, "reading the metadata file %s", metadataFilePath)
-		}
+	// Concurrency is limited via a checkout-checkin system.
+	//
+	// Each time a process starts, it checks out a hallpass from the monitor (blocking
+	// if none are available). When the process completes, it returns it's hallpass to
+	// the monitor.
+	//
+	// Because there are a fixed number of hallpasses (maxConcurrency), and each
+	// process needs a hallpass to run, we limit effective concurrency to the number
+	// of available hallpasses.
+	maxConcurrency := runtime.NumCPU()
+
+	var hallpass struct{}
+	monitor := make(chan struct{}, maxConcurrency)
+	for i := 0; i < maxConcurrency; i++ {
+		monitor <- hallpass
+	}
```

Comment:
These low-level concurrency approaches make me very nervous as, in my experience, it's very easy to unintentionally and subtly break things. Not sure on what the recommended libs are in Go world, but a quick search showed up [sourcegraph/conc](https://github.com/sourcegraph/conc) which would allow us to write this process as follows:

```go
func mapStream(
    in chan int,
    out chan int,
    f func(int) int,
) {
    s := stream.New().WithMaxGoroutines(10)
    for elem := range in {
        elem := elem
        s.Go(func() stream.Callback {
            res := f(elem)
            return func() { out <- res }
        })
    }
    s.Wait()
}
```